### PR TITLE
Purchases: Re-introduce cancelling bundled domain with plans while still in domain refund period

### PR DESCRIPTION
**Pull request by prettyboymp**
Fixes: #21628

1. Purchase a plan and a bundle domain.
2. Go to http://calypso.localhost:3000/purchases and choose that plan
3. Click "Cancel WordPress.com [Plan Type].

* You should only receive the option to cancel the domain if the domain is still within the refund window.
* The cancel button should be disabled if you have chosen to cancel the domain, but not checked the confirmation checkbox.
* After clicking the button to "Cancel Subscription and Refund", the network request should include a parameter with `cancel_bundled_domain:true` if and only if the user chose to cancel the domain and the cancel confirmation was checked.

Note, that API side changes (D9415-code) still need to be made before this can be tested end to end.

### CODE REVIEW

File: client/me/purchases/cancel-purchase/refund-information.jsx [hunk 1]
```diff
@@ -67,42 +90,130 @@ const CancelPurchaseRefundInformation = ( { purchase, includedDomainPurchase } )
...
+							/>
+							<span>
+								{ i18n.translate( 'Cancel the plan and the domain "%(domain)s."', {
+									args: {
+										domain: includedDomainPurchase.meta,
+									},
+								} ) }
+								<br />
+								{ i18n.translate(
+									"You will be refunded %(planCost)s, but the domain will be canceled and can't be " +
```
Comment (a8ci18n) [inline client/me/purchases/cancel-purchase/refund-information.jsx]: Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'You will be refunded <strong>%1$s</strong>, but the domain will be canceled and can't be guaranteed to be available for registration again.' )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/21628-allow-bundled-domain-refund)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/me/purchases/cancel-purchase/refund-information.jsx [hunk 2]
```diff
@@ -67,42 +90,130 @@ const CancelPurchaseRefundInformation = ( { purchase, includedDomainPurchase } )
...
+						),
+						<FormLabel key="keep_bundled_domain">
+							<FormRadio
+								name="keep_bundled_domain_false"
+								value="keep"
+								checked={ ! cancelBundledDomain }
+								onChange={ onCancelBundledDomainChange }
+							/>
+							<span>
+								{ i18n.translate( 'Just cancel the plan and keep %(domain)s.', {
```
Comment (a8ci18n) [inline client/me/purchases/cancel-purchase/refund-information.jsx]: Hi! I've found a possible matching string that has already been translated **22** times:
`translate( 'Just cancel the plan, and keep "%s"' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by benhuberman** [comment]
@prettyboymp  -- just to make sure I look at the right copy, is it the options shown in the second screenshot [in this issue](https://github.com/Automattic/wp-calypso/issues/21628)? Or something else?

**Comment by prettyboymp** [comment]
@benhuberman yes, it's the copy in the second "Suggested Solution" screenshot.

**Comment by prettyboymp** [comment]
cc @Automattic/payments

**Comment by benhuberman** [comment]
Gotcha, thanks @prettyboymp. I had a few suggestions for tweaks -- trying to unpack and make things a tad clearer, and a little "warmer." Instead of a running list of tweaks, here's a suggested edit of the entire chunk of copy. Let me know if you have any questions / aren't sure about any of my proposed changes:

> **Cancel your WordPress.com [Plan name] plan**
> 
> We're sorry to hear the WordPress.com [Plan name] plan didn't fit your current needs, but thank you for giving it a try.
> 
> Your plan included the custom domain [DOMAIN]. You can cancel your domain as well as the plan, but keep in mind that when you cancel a domain you risk losing it forever, and visitors to 
> your site may experience difficulties accessing it. 
> 
> We'd like to offer you two options to choose from:
> - **Cancel the plan, but keep [DOMAIN]**
> You'll receive a partial refund of [$AMOUNT] -- the cost of the [Plan name] plan, minus [$AMOUNT] for the domain. There will be no change to your domain registration, and you're free to use it on WordPress.com or transfer it elsewhere.
> 
> - **Cancel the plan **_and_** the domain [DOMAIN]**
> You'll receive a full refund of [$AMOUNT]. The domain will be canceled, and it's possible you'll lose it permanently.
> 
> When you cancel a domain, it becomes unavailable for a while. Anyone may register it once it's available again, so it's possible you won't have another chance to register it in the future. If you'd like to use your domain on a site hosted elsewhere, consider updating your name servers [LINK] instead. 
> 
> [checkbox] I understand that canceling my domain means I might never be able to register it again. 

On a sidenote, for marketing/retention reasons, I'd really want to see (or at least test) a "Recommended" label/sticker next to the first option -- I wonder if @anneforbush might have thoughts on that!

**Comment by prettyboymp** [comment]
@benhuberman Thank you for the feedback.  Here is the updated copy:

screen shot 2018-01-19 at 7 19 48 pm

**Comment by benhuberman** [comment]
Looking good!

My one last bit of feedback: the CTA button copy is slightly confusing -- people _could_ read it as "cancel the subscription **and** cancel the refund" since "cancel" is the only verb here. One way about this would be to just stick with "Cancel Subscription," and letting the refund $ indicated on the left convey the message that, yes, you're getting a refund. The other option I can think of is to tweak the language on the button -- **Cancel Plan & Receive Refund**, though it may or may not be too long for a single button. By the way, is there any particular reason we're using "subscription" here at all? The official name is the WordPress.com [Plan name] **plan**, so subscription -- both at the top and on the button -- looks somewhat inconsistent with the rest of our messaging.

**Comment by prettyboymp** [comment]
> By the way, is there any particular reason we're using "subscription" here at all?

The same code gets used by different types of products, some plans, some subscriptions.  We should probably look at creating a `genericProductName` translation property to the products/purchases in the future so we don't have to litter every place with checks against plans/domains/etc to determine what string to use.

File: client/me/purchases/cancel-purchase/index.jsx [hunk 3]
```diff
@@ -180,16 +206,26 @@ class CancelPurchase extends React.Component {
...
 }
 
-export default connect( ( state, props ) => ( {
-	hasLoadedSites: ! isRequestingSites( state ),
-	hasLoadedUserPurchasesFromServer: hasLoadedUserPurchasesFromServer( state ),
-	selectedPurchase: getByPurchaseId( state, props.purchaseId ),
-	selectedSite: getSelectedSiteSelector( state ),
-} ) )( localize( CancelPurchase ) );
+export default connect( ( state, props ) => {
+	const purchase = getByPurchaseId( state, props.purchaseId );
```
Comment (sirbrillig) [inline client/me/purchases/cancel-purchase/index.jsx]: I guess it's unlikely, but what happens if `getPurchaseById` returns `undefined`?

File: client/me/purchases/cancel-purchase/index.jsx [hunk 4]
```diff
@@ -180,16 +206,26 @@ class CancelPurchase extends React.Component {
...
+	return {
+		hasLoadedSites: ! isRequestingSites( state ),
+		hasLoadedUserPurchasesFromServer: hasLoadedUserPurchasesFromServer( state ),
+		selectedPurchase: getByPurchaseId( state, props.purchaseId ),
```
Comment (sirbrillig) [inline client/me/purchases/cancel-purchase/index.jsx]: I think this is the same as the above call and could just be replaced by `purchase` here.

File: client/me/purchases/cancel-purchase/refund-information.jsx [hunk 5]
```diff
@@ -62,43 +95,136 @@ const CancelPurchaseRefundInformation = ( { purchase, includedDomainPurchase } )
...
+							/>
+							<span>
+								{ i18n.translate( 'Cancel the plan, but keep %(domain)s.', {
+									args: {
+										domain: includedDomainPurchase.meta,
+									},
+								} ) }
+								<br />
+								{ i18n.translate(
+									"You'll receive a partial refund of %(refundAmount)s -- the cost of the %(productName)s " +
```
Comment (klimeryk) [inline client/me/purchases/cancel-purchase/refund-information.jsx]: The `--` should probably be some fancy dash - which one, I always forget 😁

File: client/me/purchases/cancel-purchase/button.jsx [hunk 6]
```diff
@@ -269,7 +270,7 @@ class CancelPurchaseButton extends Component {
 	submitCancelAndRefundPurchase = () => {
 		const { purchase, selectedSite } = this.props;
 		const refundable = isRefundable( purchase );
-
+		const cancel_bundled_domain = this.props.cancelBundledDomain;
```
Comment (klimeryk) [inline client/me/purchases/cancel-purchase/button.jsx]: We should use https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#naming-conventions code in Calypso.

File: client/me/purchases/cancel-purchase/refund-information.jsx (hunk 5 above)
Comment (Hywan) [reply client/me/purchases/cancel-purchase/refund-information.jsx]: This one: — (http://graphemica.com/%E2%80%94).

**Pull request merged**